### DISCUSSION TITLE
Protocol Gen?

### DISCUSSION THREAD

**Discussion by junebash** [stephencelis/swift-gen#26] [Ideas]
Just for fun today, I did a bit of work making a protocol version of most of this library. I made a basic benchmark using `SuspendingClock` and a moderately-nested generator, running in release mode:

```swift
let anyGen = Gen.bool
    .flatMap { $0 ? .letter : .number }
    .string(of: Gen.int(in: 10...20))
    .array(of: Gen.int(in: 5...10))
    .map { $0.joined(separator: " ") }
let protoGen = Bool.Generator()
    .flatMap { $0 ? Gens.letter : Gens.number }
    .string(of: Int.generator(in: 10...20))
    .array(of: Int.generator(in: 5...10))
    .map { $0.joined(separator: " ") }

let clock = SuspendingClock()
let repeatCount = 10_000

var rng = Xoshiro(seed: 20)

let anyTime = clock.measure {
    for _ in 1...repeatCount {
        _ = anyGen.run(using: &rng)
    }
}
let protoTime = clock.measure {
    for _ in 1...repeatCount {
        _ = protoGen.run(using: &rng)
    }
}

print("Legacy time:", anyTime)
print("Protocol time:", protoTime)
```

With the following results:
```
Legacy time: 0.21913475 seconds
Protocol time: 0.043444083 seconds
```

So the protocol version seems to perform about 5 times better, it seems. This might improve with even more nested generators.

If you're interested in perusing, [I've got the code in a fork here](https://github.com/junebash/swift-gen/tree/protocol).

**Comment by stephencelis** [comment]
Fun! Thanks for sharing the code, it's a fun read! Nice to see such performance gains, too!

Protocol-based RNGs has definitely been something we've been meaning to explore for a~ages, but other things have gotten our attention first :smile: Maybe someday! I'm also a lil curious about result builders for composing generators together, so maybe we'll try it out soon.
